### PR TITLE
Feature/66 search algorithm between two locations

### PR DESCRIPTION
**Pull request by jimibilibob**

### CODE REVIEW

File: RoutesApp_ios/utilities/extensions/CLLocationCoordinate2DExtension.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+//  CLLocationCoordinate2DExtension.swift
+//  RoutesApp_ios
+//
+//  Created by user on 29/8/22.
+//
+
+import CoreLocation
+
+extension CLLocationCoordinate2D {
+    /// Returns the distance between two coordinates in meters.
```
Comment (jmacboy) [inline RoutesApp_ios/utilities/extensions/CLLocationCoordinate2DExtension.swift]: you have an extra /

File: RoutesApp_ios/utilities/helpers/Algorithm.swift [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                for stop in routeFromDestination.stops {
+                    let nearestStopToOrigin = routeFromOrigin.stops.min(by: {
+                        $0.toCLLocationCoordinate2D().distance(to: stop.toCLLocationCoordinate2D()) <=
+                            $1.toCLLocationCoordinate2D().distance(to: stop.toCLLocationCoordinate2D())
+                    })
+                    guard let nearestStopToOrigin = nearestStopToOrigin else { continue }
+                    let indexOfNearestStopToOrigin = getIndexWhere(coordinate: nearestStopToOrigin, coordinateList: routeFromOrigin.stops)
+                    if nearestStopToOrigin.toCLLocationCoordinate2D().distance(to: stop.toCLLocationCoordinate2D()) <= minDistanceBtwStops {
+                        // MARK: Origin line splice
+                        let indexOfOriginRoute = getIndexWhere(coordinate: nearestStopToOrigin, coordinateList: routeFromOrigin.routePoints)
```
Comment (jmacboy) [inline RoutesApp_ios/utilities/helpers/Algorithm.swift]: The other methods are great but this one is a little bit bigger.
Try to reduce cognitive complexity inside this if, call a couple of methods to obtain the values that you´re calculating here.

Comment (jmacboy) [reply RoutesApp_ios/utilities/helpers/Algorithm.swift]: as a hint, every MARK you´re using can be separated as a method

File: RoutesApp_iosTests/AlgorithmTests.swift [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            XCTAssertTrue(false)
+            return
+        }
+
+        let result = Algorithm.shared.findAvailableRoutes(origin: originPoint, destination: destinationPoint,
+            lines: [line1, line2], minDistanceBtwPoints: minDistance, minDistanceBtwStops: minDistanceBtwStops)
+        XCTAssertNotNil(result[0].connectionPoint)
+    }
+
+    func testPerformanceExample() throws {
```
Comment (jmacboy) [inline RoutesApp_iosTests/AlgorithmTests.swift]: No need to add this if you´re not going to use it

File: RoutesApp_iosTests/AlgorithmTests.swift [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        line1 = Line(name: "01", categoryRef: "LineCategories/DW8blCpvs0OXwkozaEhn",
+                     routePoints: routePoints1, start: routePoints1[0], stops: stops1)
+        // MARK: Route 2
+        let routePoints2 = TestResources.points2Array
+        let stops2 = TestResources.stops2Array
+        line2 = Line(name: "1001", categoryRef: "LineCategories/JY8blWsvs0OXwkozaJlb",
+                     routePoints: routePoints2, start: routePoints2[0], stops: stops2)
+    }
+
+    override func tearDownWithError() throws {
```
Comment (jmacboy) [inline RoutesApp_iosTests/AlgorithmTests.swift]: No need to add this if you´re not going to use it

File: RoutesApp_iosTests/AlgorithmTests.swift [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            XCTAssertTrue(false)
+            return
+        }
+
+        let result = Algorithm.shared.findAvailableRoutes(origin: originPoint, destination: destinationPoint,
+            lines: [line1, line2], minDistanceBtwPoints: minDistance, minDistanceBtwStops: minDistanceBtwStops)
+        XCTAssertGreaterThan(result.count, 0)
+    }
+
+    func testWhenGivenLinesAndNeedToCombineRoutes() throws {
```
Comment (jmacboy) [inline RoutesApp_iosTests/AlgorithmTests.swift]: Add a couple of route examples more to your tests, one that only has double routes, one that only has single routes, one that has both (I think the last one is covered), also one that doesn't have any route

**Review by jmacboy — CHANGES_REQUESTED** [review]

File: RoutesApp_ios/utilities/extensions/CLLocationCoordinate2DExtension.swift (hunk 1 above)
Comment (jimibilibob) [reply RoutesApp_ios/utilities/extensions/CLLocationCoordinate2DExtension.swift]: Removed extra /

File: RoutesApp_ios/utilities/helpers/Algorithm.swift (hunk 2 above)
Comment (jimibilibob) [reply RoutesApp_ios/utilities/helpers/Algorithm.swift]: Added more methods to make it easier to understand

File: RoutesApp_iosTests/AlgorithmTests.swift (hunk 3 above)
Comment (jimibilibob) [reply RoutesApp_iosTests/AlgorithmTests.swift]: Removed unnecessary tests

File: RoutesApp_iosTests/AlgorithmTests.swift (hunk 4 above)
Comment (jimibilibob) [reply RoutesApp_iosTests/AlgorithmTests.swift]: Removed unnecessary test

File: RoutesApp_iosTests/AlgorithmTests.swift (hunk 5 above)
Comment (jimibilibob) [reply RoutesApp_iosTests/AlgorithmTests.swift]: Added a case that on has a double route and another one that just have a single route.

File: RoutesApp_iosTests/AlgorithmTests.swift [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+        guard let line1 = line1,
+                let line2 = line2 else {
+            XCTAssertTrue(false)
+            return
+        }
+
+        let result = Algorithm.shared.findAvailableRoutes(origin: originPoint, destination: destinationPoint,
+            lines: [line1, line2], minDistanceBtwPoints: minDistance, minDistanceBtwStops: minDistanceBtwStops)
+        XCTAssertGreaterThan(result.count, 0)
```
Comment (jmacboy) [inline RoutesApp_iosTests/AlgorithmTests.swift]: you should not only check the result count, you should also check that the returned result is what you expect it to be

File: RoutesApp_iosTests/AlgorithmTests.swift [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+        guard let line1 = line1,
+                let line2 = line2 else {
+            XCTAssertTrue(false)
+            return
+        }
+
+        let result = Algorithm.shared.findAvailableRoutes(origin: originPoint, destination: destinationPoint,
+            lines: [line1, line2], minDistanceBtwPoints: minDistance, minDistanceBtwStops: minDistanceBtwStops)
+        XCTAssertNotNil(result[0].connectionPoint)
```
Comment (jmacboy) [inline RoutesApp_iosTests/AlgorithmTests.swift]: you should not only check the connection point, you should also check that the returned result is what you expect it to be

File: RoutesApp_iosTests/AlgorithmTests.swift [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
```
Comment (jmacboy) [inline RoutesApp_iosTests/AlgorithmTests.swift]: you should not only check the connection point, you should also check that the returned result is what you expect it to be

**Review by jmacboy — CHANGES_REQUESTED** [review]

File: RoutesApp_iosTests/AlgorithmTests.swift (hunk 6 above)
Comment (jimibilibob) [reply RoutesApp_iosTests/AlgorithmTests.swift]: Enhanced

File: RoutesApp_iosTests/AlgorithmTests.swift (hunk 7 above)
Comment (jimibilibob) [reply RoutesApp_iosTests/AlgorithmTests.swift]: Checking the the returned result

File: RoutesApp_iosTests/AlgorithmTests.swift (hunk 8 above)
Comment (jimibilibob) [reply RoutesApp_iosTests/AlgorithmTests.swift]: Checking the object result

**Review by jmacboy — APPROVED** [review]

**Review by adCBueno — APPROVED** [review]

**Pull request merged**